source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,296 @@
+    const protobufs::InstructionDescriptor& atomic_instruction,
+    uint32_t memory_semantics_operand_index,
+    uint32_t memory_semantics_new_value_id) {
+  *message_.mutable_atomic_instruction() = atomic_instruction;
+
+  assert(IsNeededOpcodeWithAppropriateIndex(
+             static_cast<SpvOp>(atomic_instruction.target_instruction_opcode()),
+             memory_semantics_operand_index) &&
+         "The instruction may not be an atomic instruction, or an operand "
+         "index does not equal 0 or 1.");
+
+  message_.set_memory_semantics_operand_index(memory_semantics_operand_index);
+  message_.set_memory_semantics_new_value_id(memory_semantics_new_value_id);
+}
+
+bool TransformationChangingMemorySemantics::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  // The atomic instruction must exist.
+  if (!needed_atomic_instruction) {
+    return false;
+  }
+  // Instruction must be atomic instruction only. Memory semantics operand Index
+  // must be equal to 0 or 1 only. Operand index must be equal to 0 in the case
+  // of instruction that takes one memory semantics operand and 0 or 1 for
+  // instructions that takes two operands.
+  if (!IsNeededOpcodeWithAppropriateIndex(
+          needed_atomic_instruction->opcode(),
+          message_.memory_semantics_operand_index())) {
+    return false;
+  }
+
+  // The new value instruction must exist and must be OpConstant.
+  auto value_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!value_instruction) {
+    return false;
+  }
+  if (value_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+
+  // The first 5 bits of new memory semantics values must be suitable for needed
+  // instruction. The first 5 bits of new value must be larger than the first 5
+  // bits of old. Can't use Sequentially Consistent memory semantic if the
+  // memory model is Vulkan.
+  auto new_memory_sematics_value = value_instruction->GetSingleWordInOperand(0);
+  uint32_t old_memory_sematics_value = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    // Memory semantics true index for the OpMemoryBarrier equal 1.
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(1))
+            ->GetSingleWordInOperand(0);
+  } else {
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                message_.memory_semantics_operand_index() == 0 ? 2 : 3))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics =
+      static_cast<SpvMemorySemanticsMask>(new_memory_sematics_value & 0x1F);
```

Comment:
I suggest you extract all hex values into global constants. E.g. `kMemorySemanticsHigherBitmask`. [Like this](https://github.com/KhronosGroup/SPIRV-Tools/blob/bed84792f9dd77bca4fefb72195be153bc64fa4b/source/fuzz/transformation_replace_load_store_with_copy_memory.cpp#L25). And add comments above each constant to explain what it is for.

Reply:
Done.